File: platform/lib/domain/src/main/kotlin/io/hamal/lib/domain/request/Namespace.kt
package io.hamal.lib.domain.request

import io.hamal.lib.domain._enum.RequestStatus
import io.hamal.lib.domain.vo.*

interface NamespaceAppendRequest {
    val name: NamespaceName
    val features: NamespaceFeaturesMap?
}

data class NamespaceAppendRequested(
    override val requestId: RequestId,
    override val requestedBy: AuthId,
    override var requestStatus: RequestStatus,
    val id: NamespaceId,
    val workspaceId: WorkspaceId,
    val parentId: NamespaceId,
    val name: NamespaceName,
    val features: NamespaceFeaturesMap?
) : Requested()

interface NamespaceUpdateRequest {
    val name: NamespaceName?
    val features: NamespaceFeaturesMap?
}

data class NamespaceUpdateRequested(
    override val requestId: RequestId,
    override val requestedBy: AuthId,
    override var requestStatus: RequestStatus,
    val id: NamespaceId,
    val workspaceId: WorkspaceId,
    val name: NamespaceName?,
    val features: NamespaceFeaturesMap?
) : Requested()

File: app/fn-guru/src/main/frontend/src/types/namespace.ts


export interface NamespaceAppendRequested {
    requestId: string;
    requestStatus: string;
    id: string;
    workspaceId: string;
}

export interface NamespaceUpdateRequested {
    requestId: string;
    requestStatus: string;
    id: string;
}

export interface NamespaceFeature {
    schedule : boolean
    topic : boolean
    webhook : boolean
    endpoint : boolean
}

export interface Namespace {
    id: string;
    name: string;
    features: NamespaceFeature;
}

export interface NamespaceList {
    namespaces: Array<NamespaceListItem>;
}

export interface NamespaceListItem {
    id: string;
    parentId: string;
    name: string;
}

export class NamespaceNode {
    data: NamespaceListItem
    descendants: Array<NamespaceNode>

    constructor(value: NamespaceListItem) {
        this.data = value
        this.descendants = []
    }

    addDescendant(other: NamespaceNode) {
        if (this.data.id !== other.data.id) {
            this.descendants.push(other)
        }
    }

    isParent() {
        return this.descendants.length > 0
    }
}


